server/src/test/java/org/opensearch/index/shard/RemoteIndexShardTests.java

```diff
@@ -343,6 +360,121 @@ public void testPrimaryRestart() throws Exception {
         }
     }
 
+    /**
+     * This test validates that unreferenced on disk file are ignored while requesting files from replication source to
+     * prevent FileAlreadyExistsException. It does so by only copying files in first round of segment replication without
+     * committing locally so that in next round of segment replication those files are not considered for download again
+     */
+    public void testSegRepSucceedsOnPreviousCopiedFiles() throws Exception {
```

Comment:
This test is pretty complex and makes me worried about flakiness going fwd with all the async behavior.

The getFiles method itself does not make any async call, its only validating the local dir against the retrieved metadata.  Can we make this package private / visible for testing and unit test it directly?  This would greatly simplify this and only require mocking the checkpoint response and validate against a replica's local store.

Reply:
>This test is pretty complex and makes me worried about flakiness going fwd with all the async behavior.

Ran test locally 100 times without failure. 

>The getFiles method itself does not make any async call, its only validating the local dir against the retrieved metadata. Can we make this package private / visible for testing and unit test it directly? This would greatly simplify this and only require mocking the checkpoint response and validate against a replica's local store.

IMO, this test provides top level coverage which validates that when store has unreferenced files, then follow up round of segment replication does not fail. Though unit tests does provide similar coverage, but I prefer shard level tests as it gives more confidence.